honour routes timeout if max stream duration is set with out stream duration

ramaraochavali opened the pull request:
Commit Message: honour routes timeout if maxstream duration is set with out specifying max stream duration inside it.
Additional Description:
Risk Level: Low
Testing: Added
Docs Changes: N/A
Release Notes: N/A (There is already a bug fix mentioned for MaxStreamDuration in current.rst)
Platform Specific Features: N/A
Fixes https://github.com/envoyproxy/envoy/issues/15530

ramaraochavali commented:
@mattklein123 @markdroth PTAL

jmarantz commented:
@adamsjob FYI

mattklein123 commented:
Assigning over to @alyssawilk who I think reviewed the other timeout PRs.

alyssawilk commented on source/common/router/router.cc:
```diff
@@ -156,6 +156,13 @@ FilterUtility::finalTimeout(const RouteEntry& route, Http::RequestHeaderMap& req
     } else {
       timeout.global_timeout_ = route.timeout();
     }
+  } else {
```
When we switched timeouts, we moved the instantiation of the timeout to
ConnectionManagerImpl::ActiveStream::refreshDurationTimeout()

Mind moving this logic there?

alyssawilk reviewed (commented):
Thanks for catching and fixing this!

ramaraochavali commented on test/common/http/conn_manager_impl_test.cc:
```diff
@@ -2048,6 +2048,61 @@ TEST_F(HttpConnectionManagerImplTest, TestStreamIdleAccessLog) {
+// Test timeout when route timeout needs to be used if max_stream_duration is not specified.
+// Regression test for https://github.com/envoyproxy/envoy/issues/15530.
+// TODO(ramaraochavali): move this test to DurationTimeout test.
+TEST_F(HttpConnectionManagerImplTest, DurationTimeoutFromRouteTimeout) {
```
I had to create a separate test for this because `clusterInfo` is repeatedly called and it keeps triggering the clearRouteCache so the timer gets fired multiple times. The same problem exists for https://github.com/envoyproxy/envoy/blob/main/test/common/http/conn_manager_impl_test.cc#L2173 where timer is repeatedly enabled. Having two such tests in the same test is not working. So I made it as separate test and TODO to look at the issue later.

ramaraochavali commented:
@alyssawilk moved to `refreshDurationTimeout`. PTAL.

markdroth commented:
Alyssa knows this code a lot better than I do, so I'll let her review.

ramaraochavali commented:
@alyssawilk can you PTAL when you get chance?

alyssawilk commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -1176,8 +1176,14 @@ void ConnectionManagerImpl::ActiveStream::refreshDurationTimeout() {
+    // If we are using new timeouts and MaxStreamDuration is not set, use the route's timeout
+    // value. This is the case when route's MaxStreamDuration is specified without inner
+    // MaxStreamDuration i.e. only grpc_timeout_header_max is set.
+    if (!route->maxStreamDuration()) {
```
so if grpc_timeout_header_max is set when do we use it?  usingNewTimeouts would be true and we'd ignore all the code below, no?

Can you please fix and regression test?

ramaraochavali replied:
Oh, Yeah.. Missed that. Will add a test

ramaraochavali replied:
@alyssawilk Can you PTAL at the fix to see if this is correct? One thing I found was this needs changes to most of the test cases because now we will be initiating request timer most of the times. So I just want to ensure this is the right direction before I make changes the tests(some of which I have already done - please take a look at the changes)

alyssawilk commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -1173,45 +1173,66 @@ void ConnectionManagerImpl::ActiveStream::refreshDurationTimeout() {
-  auto grpc_timeout = Grpc::Common::getGrpcTimeout(*request_headers_);
   std::chrono::milliseconds timeout;
   bool disable_timer = false;
+  auto grpc_timeout = Grpc::Common::getGrpcTimeout(*request_headers_);
```
Offhand I recall debate around the original PR about how the new fields should be applied and it was the consensus from the gRPC folks that the difference wasn't "is this a gRPC request or not" but "is the gRPC timeout header present or not" so I think switching to this is against the intent of the folks who added this to the API.

ramaraochavali replied:
Got it. This is good to know. I will change it accordingly

alyssawilk replied:
Awesome.  It's super complicated and confusing, so thanks for continuing to plug away at it!

ramaraochavali commented:
@alyssawilk Sorry for the delay. Fixed. PTAL. I have added a regression test for this and left a TODO to merge to DurationTimeout test later as I am having some issues with mock timer (even with existing tests)

alyssawilk commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -1192,6 +1192,9 @@ void ConnectionManagerImpl::ActiveStream::refreshDurationTimeout() {
         timeout = max_stream_duration.value();
+      } else if (route->timeout().count() != 0) {
+        // If max stream duration is not set either at route/HCM level, use the route timeout.
+        timeout = route->timeout();
```
Yeah, I think think this is right.
I'm going to tag @antoniovicente for second pass of code as a second eye to make sure we got the various combinations and permutations right, and I'll do a second pass when you've got the tests sorted out.

antoniovicente replied:
I'm wondering if something else in the system is already enforcing the relevant timeout.  I found use of timeout_.global_timeout_ in Filter::onRequestComplete, but I think that timer is not comprehensive as it waits until the full request body has been received before it is activated.  I wonder if we could use the same logic to compute the timeout in both places and/or have a single timer to control this behavior.

https://github.com/envoyproxy/envoy/blob/0314715ee7cfaa72b57aa592b13d89f4b90cb9c1/source/common/router/router.cc#L830

ramaraochavali replied:
AFAIK it is difficult to unify that logic as that one is meant with old timeouts. @alyssawilk any thoughts?

alyssawilk replied:
Yeah, I think unifying is outside the scope of this PR so I'd be inclined to just land the simpler fix.

antoniovicente replied:
Ok.  I think that the timeout handling added in this PR is fine.

antoniovicente reviewed (approved):
Thanks for the fix to timeout config handling.

ramaraochavali commented:
@alyssawilk Can you PTAL when you get chance?

alyssawilk reviewed (approved):
Ah sorry, I thought you meant to address the test TODO in this PR.  I'm fine with it being a follow-up!

rgs1 commented:
FYI this broke web sockets for us (by forcing them to terminate earlier), wondering if we can contribute more integration tests here to catch these sort of regressions.
